src/instruction.rs

```diff
@@ -0,0 +1,195 @@
+//! Program instruction types.
+
+use solana_program::{
+    instruction::{AccountMeta, Instruction},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+    system_program,
+};
+
+/// Instructions supported by the Paladin Rewards program.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum PaladinRewardsInstruction {
+    /// Configures staker rewards for a mint that has been configured with the
+    /// rewards program as a transfer hook program.
+    ///
+    /// This instruction will:
+    ///
+    /// - Initialize a staker rewards account.
+    /// - Initialize the required accounts for the transfer hook.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[w]` Staker rewards account.
+    /// 1. `[w]` Transfer hook extra account metas account.
+    /// 2. `[ ]` Token mint.
+    /// 3. `[ ]` Token mint.
+    /// 4. `[s]` Mint authority.
+    /// 5. `[ ]` System program.
+    InitializeStakerRewards,
+    /// Moves SOL rewards to the following parties:
+    ///
+    /// - 1%  Piggy bank.
+    /// - 4%  Staker rewards.
+    /// - 5%  Holder rewards.
+    /// - 90% Leader who produces the block.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[w]` Staker rewards account.
+    /// 1. `[w]` Holder rewards account.
+    /// 2. `[w]` Piggy bank account.
+    /// 3. `[w]` Leader account.
+    DistributeRewards,
```

Comment:
Based on our discussion, this should probably only take the payer / leader account, total rewards account, and the system program, right?

Reply:
This instruction is for the leader to distribute rewards to the holder rewards pool, right? If so, should the leader account be a signer, too?

Reply:
Hm... that's a good question, I'm not sure to be honest. I was thinking of keeping this simple and just having a payer sign to transfer lamports into the holder rewards pool, and not necessarily needing a separate leader account.

Let's see... we'll be able to look off-chain to see how the payer got those funds during their leader slot, and in fact, we'll need to do that to make sure that validators are acting honestly and paying out the right amount. Validators running the bot are expected to pay out rewards during their slot, and if they *don't*, they'll run the risk of getting slashed by the network. There's no incentive for someone else (ie. not a validator running the bot) to pay out rewards, so I think we're ok with omitting validator info.

What do you think?

Reply:
Ok great, yeah that works for me. I just have a slight preference for reversing so that the order goes source then destination:

```
/// Moves SOL rewards to the holder rewards pool and updates the total.
///
/// Accounts expected by this instruction:
///
/// 0. `[w, s]` Payer account.
/// 1. `[w]` Holder rewards pool account.
/// 2. `[ ]` System program.
DistributeRewards(u64),
```

Reply:
Not a hill I'd die on, but the rest of my instructions go like this:

0. Holder rewards pool
1. Holder rewards
2. Token account
3. Token mint
4. Signer
5. Programs ...